py: add ruff rules

isinyaaa opened the pull request:
As suggested by @dtrifiro, let's add some Ruff rules to the python client config.

## Description
Add relevant rules to keep python code in check. See https://github.com/isinyaaa/model-registry/blob/878aa46755cc3a0852f792d972b10be0dadc753b/clients/python/pyproject.toml for more details on the rules enabled.

Most new warnings have been fixed as well.

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

dtrifiro commented on clients/python/pyproject.toml:
```diff
@@ -39,7 +39,34 @@ line-length = 119
+]
+ignore = [
+    "E501", # line too long
+    "S101", # use of assert detected
```
Not sure if you use this for non-test code, but you can add a per-file-ignore like so:

```toml
[tool.ruff.per-file-ignores]
"__init__.py" = ["F401"]
"tests/*" = ["S101"]
"tests/conftest.py" = ["F403"]
```

dtrifiro replied:
Not sure if you want to add an exception for `D102 Missing docstring in public method` somewhere. CI is failing due to that

isinyaaa replied:
Resolved using `typing_extensions` to add `@override`.

dtrifiro commented on clients/python/tests/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# ruff: noqa: D101
```
this could also be added to the per-files-ignore list

isinyaaa replied:
I'll check that option.

isinyaaa commented:
Rebased to add two commits that I had planned for later. I squashed the current [2/6] with the previous [2/5]. I believe this makes some annotations easier and also simplifies quite nicely. I still haven't dealt with all of mypy's complaints though, and I see it isn't running on the CI.

tarilabs reviewed (approved):
reconfirming my previous approval [here](https://github.com/opendatahub-io/model-registry/pull/181#event-11050421696) 👍

tarilabs reviewed (approved):
- keep https://github.com/kubeflow/model-registry/pull/858
- keep https://github.com/kubeflow/model-registry/pull/859
- keep https://github.com/kubeflow/model-registry/pull/860
- keep https://github.com/kubeflow/model-registry/pull/861
- keep https://github.com/kubeflow/model-registry/pull/838

- keep https://github.com/kubeflow/model-registry/pull/820
technically this is not really required (we don't push UI images midstream), but since we're just using to verify integrity of the docker build, we can keep like for others. we can revisit as needed

- keep https://github.com/kubeflow/model-registry/pull/852
- keep https://github.com/kubeflow/model-registry/pull/855

labels are applied

/lgtm
/approve
